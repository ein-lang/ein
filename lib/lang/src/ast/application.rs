use super::expression::Expression;
use crate::{
    debug::SourceInformation,
    types::{self, Type},
};
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct Application {
    type_: Type,
    function: Arc<Expression>,
    argument: Arc<Expression>,
    source_information: Arc<SourceInformation>,
}

impl Application {
    pub fn new(
        function: impl Into<Expression>,
        argument: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        let source_information = source_information.into();

        Self::with_type(
            types::Unknown::new(source_information.clone()),
            function,
            argument,
            source_information,
        )
    }

    pub fn with_type(
        type_: impl Into<Type>,
        function: impl Into<Expression>,
        argument: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        Self {
            type_: type_.into(),
            function: Arc::new(function.into()),
            argument: Arc::new(argument.into()),
            source_information: source_information.into(),
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn function(&self) -> &Expression {
        &self.function
    }

    pub fn argument(&self) -> &Expression {
        &self.argument
    }

    pub fn source_information(&self) -> &Arc<SourceInformation> {
        &self.source_information
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::with_type(
            self.type_.clone(),
            self.function.transform_expressions(transform)?,
            self.argument.transform_expressions(transform)?,
            self.source_information.clone(),
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::with_type(
            self.type_.transform_types(transform)?,
            self.function.transform_types(transform)?,
            self.argument.transform_types(transform)?,
            self.source_information.clone(),
        ))
    }
}
