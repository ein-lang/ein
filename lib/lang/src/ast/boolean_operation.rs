use super::expression::Expression;
use crate::{debug::SourceInformation, types::Type};
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum BooleanOperator {
    And,
    Or,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BooleanOperation {
    operator: BooleanOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    source_information: Arc<SourceInformation>,
}

impl BooleanOperation {
    pub fn new(
        operator: BooleanOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        Self {
            operator,
            lhs: lhs.into().into(),
            rhs: rhs.into().into(),
            source_information: source_information.into(),
        }
    }

    pub fn operator(&self) -> BooleanOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn source_information(&self) -> &Arc<SourceInformation> {
        &self.source_information
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.operator,
            self.lhs.transform_expressions(transform)?,
            self.rhs.transform_expressions(transform)?,
            self.source_information.clone(),
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.operator,
            self.lhs.transform_types(transform)?,
            self.rhs.transform_types(transform)?,
            self.source_information.clone(),
        ))
    }
}
