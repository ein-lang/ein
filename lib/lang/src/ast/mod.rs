mod alternative;
mod application;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod case;
mod definition;
mod equality_operation;
mod export;
mod expression;
mod function_definition;
mod if_;
mod import;
mod let_;
mod list;
mod list_case;
mod list_element;
mod module;
mod module_interface;
mod none;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_element_operation;
mod record_update;
mod string;
mod type_coercion;
mod type_definition;
mod unresolved_import;
mod unresolved_module;
mod variable;
mod variable_definition;

pub use alternative::*;
pub use application::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use case::*;
pub use definition::*;
pub use equality_operation::*;
pub use export::*;
pub use expression::*;
pub use function_definition::*;
pub use if_::*;
pub use import::*;
pub use let_::*;
pub use list::*;
pub use list_case::*;
pub use list_element::*;
pub use module::*;
pub use module_interface::*;
pub use none::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_element_operation::*;
pub use record_update::*;
pub use string::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use unresolved_import::*;
pub use unresolved_module::*;
pub use variable::*;
pub use variable_definition::*;
