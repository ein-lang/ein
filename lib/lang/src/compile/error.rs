use crate::{ast, debug::*, path::ModulePath, types};
use std::{error::Error, fmt::Display, sync::Arc};

#[derive(Debug, PartialEq)]
pub enum CompileError {
    AnyEqualOperation(Arc<SourceInformation>),
    CaseArgumentTypeInvalid(Arc<SourceInformation>),
    DuplicateNames(Arc<SourceInformation>, Arc<SourceInformation>),
    ExportedNameNotFound {
        name: String,
    },
    FunctionEqualOperation(Arc<SourceInformation>),
    FunctionExpected(Arc<SourceInformation>),
    MainFunctionNotFound(ModulePath),
    RecordElementNotFound {
        record_type: types::Record,
        name: String,
    },
    RecordEqualOperation(Arc<SourceInformation>),
    EirFmmCompile(eir_fmm::CompileError),
    TypeNotFound(types::Reference),
    TypesNotMatched(Arc<SourceInformation>, Arc<SourceInformation>),
    TypeNotInferred(Arc<SourceInformation>),
    VariableNotFound(ast::Variable),
}

impl Display for CompileError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::AnyEqualOperation(source_information) => write!(
                formatter,
                "cannot compare Any type values\n{}",
                source_information
            ),
            Self::CaseArgumentTypeInvalid(source_information) => write!(
                formatter,
                "invalid argument type of case expression\n{}",
                source_information
            ),
            Self::ExportedNameNotFound { name } => {
                write!(formatter, "exported name \"{}\" not found", name)
            }
            Self::DuplicateNames(one, other) => {
                write!(formatter, "duplicate names\n{}\n{}", one, other)
            }
            Self::FunctionEqualOperation(source_information) => write!(
                formatter,
                "cannot compare functions\n{}",
                source_information
            ),
            Self::FunctionExpected(source_information) => {
                write!(formatter, "function expected\n{}", source_information)
            }
            Self::MainFunctionNotFound(path) => write!(
                formatter,
                "main function not found in main module {}",
                &path
            ),
            Self::RecordElementNotFound { record_type, name } => write!(
                formatter,
                "element \"{}\" not found in record type\n{}",
                &name,
                record_type.source_information()
            ),
            Self::RecordEqualOperation(source_information) => write!(
                formatter,
                "cannot compare records including functions or Any values\n{}",
                source_information
            ),
            Self::EirFmmCompile(error) => {
                write!(formatter, "failed to compile eir to fmm: {:?}", error)
            }
            Self::TypeNotFound(reference) => write!(
                formatter,
                "type \"{}\" not found\n{}",
                reference.name(),
                reference.source_information()
            ),
            Self::TypeNotInferred(source_information) => {
                write!(formatter, "failed to infer type\n{}", source_information)
            }
            Self::TypesNotMatched(lhs_source_information, rhs_source_information) => write!(
                formatter,
                "types not matched\n{}\n{}",
                lhs_source_information, rhs_source_information
            ),
            Self::VariableNotFound(variable) => write!(
                formatter,
                "variable \"{}\" not found\n{}",
                variable.name(),
                variable.source_information()
            ),
        }
    }
}

impl Error for CompileError {}

impl From<eir_fmm::CompileError> for CompileError {
    fn from(error: eir_fmm::CompileError) -> Self {
        Self::EirFmmCompile(error)
    }
}
