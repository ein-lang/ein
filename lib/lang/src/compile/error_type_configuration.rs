#[cfg(test)]
use once_cell::sync::Lazy;
use std::collections::HashMap;
#[cfg(test)]
use std::sync::Arc;

#[cfg(test)]
pub static ERROR_TYPE_CONFIGURATION: Lazy<Arc<ErrorTypeConfiguration>> = Lazy::new(|| {
    ErrorTypeConfiguration {
        error_type_name: "Error".into(),
    }
    .into()
});

pub struct ErrorTypeConfiguration {
    pub error_type_name: String,
}

impl ErrorTypeConfiguration {
    pub fn qualify(&self, names: &HashMap<String, String>) -> Self {
        Self {
            error_type_name: self.qualify_name(&self.error_type_name, names),
        }
    }

    fn qualify_name(&self, name: &str, names: &HashMap<String, String>) -> String {
        names.get(name).cloned().unwrap_or_else(|| name.into())
    }
}
