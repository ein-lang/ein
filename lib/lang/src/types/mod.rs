mod any;
mod boolean;
mod function;
mod list;
mod none;
mod number;
mod record;
mod record_element;
mod reference;
mod string;
mod type_;
mod union;
mod unknown;
mod variable;

pub use any::*;
pub use boolean::*;
pub use function::*;
pub use list::*;
pub use none::*;
pub use number::*;
pub use record::*;
pub use record_element::*;
pub use reference::*;
pub use string::*;
pub use type_::*;
pub use union::*;
pub use unknown::*;
pub use variable::*;
