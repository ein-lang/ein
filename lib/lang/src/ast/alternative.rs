use super::expression::Expression;
use crate::types::Type;

#[derive(Clone, Debug, PartialEq)]
pub struct Alternative {
    type_: Type,
    expression: Expression,
}

impl Alternative {
    pub fn new(type_: impl Into<Type>, expression: impl Into<Expression>) -> Self {
        Self {
            type_: type_.into(),
            expression: expression.into(),
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn expression(&self) -> &Expression {
        &self.expression
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.type_.clone(),
            self.expression.transform_expressions(transform)?,
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.type_.transform_types(transform)?,
            self.expression.transform_types(transform)?,
        ))
    }
}
