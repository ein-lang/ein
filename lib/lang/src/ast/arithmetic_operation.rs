use super::expression::Expression;
use crate::debug::SourceInformation;
use crate::types::Type;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum ArithmeticOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
}

impl From<ArithmeticOperator> for ssf::ir::PrimitiveOperator {
    fn from(operator: ArithmeticOperator) -> Self {
        match operator {
            ArithmeticOperator::Add => ssf::ir::PrimitiveOperator::Add,
            ArithmeticOperator::Subtract => ssf::ir::PrimitiveOperator::Subtract,
            ArithmeticOperator::Multiply => ssf::ir::PrimitiveOperator::Multiply,
            ArithmeticOperator::Divide => ssf::ir::PrimitiveOperator::Divide,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArithmeticOperation {
    operator: ArithmeticOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    source_information: Arc<SourceInformation>,
}

impl ArithmeticOperation {
    pub fn new(
        operator: ArithmeticOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        Self {
            operator,
            lhs: lhs.into().into(),
            rhs: rhs.into().into(),
            source_information: source_information.into(),
        }
    }

    pub fn operator(&self) -> ArithmeticOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn source_information(&self) -> &Arc<SourceInformation> {
        &self.source_information
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.operator,
            self.lhs.transform_expressions(transform)?,
            self.rhs.transform_expressions(transform)?,
            self.source_information.clone(),
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.operator,
            self.lhs.transform_types(transform)?,
            self.rhs.transform_types(transform)?,
            self.source_information.clone(),
        ))
    }
}
