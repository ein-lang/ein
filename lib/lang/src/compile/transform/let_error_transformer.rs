use super::super::error::CompileError;
use super::super::error_type_configuration::ErrorTypeConfiguration;
use super::super::reference_type_resolver::ReferenceTypeResolver;
use super::super::type_canonicalizer::TypeCanonicalizer;
use super::super::type_equality_checker::TypeEqualityChecker;
use crate::ast::*;
use crate::debug::SourceInformation;
use crate::types::{self, Type};
use std::sync::Arc;

pub struct LetErrorTransformer {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    type_equality_checker: Arc<TypeEqualityChecker>,
    type_canonicalizer: Arc<TypeCanonicalizer>,
    error_type_configuration: Arc<ErrorTypeConfiguration>,
}

impl LetErrorTransformer {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        type_equality_checker: Arc<TypeEqualityChecker>,
        type_canonicalizer: Arc<TypeCanonicalizer>,
        error_type_configuration: Arc<ErrorTypeConfiguration>,
    ) -> Arc<Self> {
        Self {
            reference_type_resolver,
            type_equality_checker,
            type_canonicalizer,
            error_type_configuration,
        }
        .into()
    }

    pub fn transform(&self, let_: &LetError) -> Result<Expression, CompileError> {
        let error_type = types::Reference::new(
            &self.error_type_configuration.error_type_name,
            let_.source_information().clone(),
        );

        let_.definitions().iter().rev().fold(
            Ok(let_.expression().clone()),
            |expression, variable_definition| {
                let ok_type = variable_definition.type_().clone();

                Ok(Case::with_type(
                    self.type_canonicalizer.canonicalize(
                        &types::Union::new(
                            vec![ok_type.clone(), error_type.clone().into()],
                            let_.source_information().clone(),
                        )
                        .into(),
                    )?,
                    variable_definition.name(),
                    variable_definition.body().clone(),
                    vec![
                        Alternative::new(ok_type, expression?),
                        Alternative::new(
                            error_type.clone(),
                            self.coerce_type(
                                &Variable::new(
                                    variable_definition.name(),
                                    let_.source_information().clone(),
                                )
                                .into(),
                                &error_type.clone().into(),
                                &self.type_canonicalizer.canonicalize(
                                    &types::Union::new(
                                        vec![let_.type_().clone(), error_type.clone().into()],
                                        let_.source_information().clone(),
                                    )
                                    .into(),
                                )?,
                                let_.source_information().clone(),
                            )?,
                        ),
                    ],
                    let_.source_information().clone(),
                )
                .into())
            },
        )
    }

    fn coerce_type(
        &self,
        expression: &Expression,
        from_type: &Type,
        to_type: &Type,
        source_information: Arc<SourceInformation>,
    ) -> Result<Expression, CompileError> {
        Ok(
            if self.type_equality_checker.equal(&from_type, &to_type)?
                || self.reference_type_resolver.is_list(&from_type)?
                    && self.reference_type_resolver.is_list(&to_type)?
            {
                expression.clone()
            } else {
                TypeCoercion::new(
                    expression.clone(),
                    from_type.clone(),
                    to_type.clone(),
                    source_information,
                )
                .into()
            },
        )
    }
}

#[cfg(test)]
mod tests {
    use super::super::super::error_type_configuration::ERROR_TYPE_CONFIGURATION;
    use super::*;
    use crate::debug::*;
    use crate::package::Package;
    use crate::path::ModulePath;
    use lazy_static::lazy_static;

    lazy_static! {
        static ref MODULE: Module = Module::new(
            ModulePath::new(Package::new("", ""), vec![]),
            Export::new(Default::default()),
            vec![Import::new(
                ModuleInterface::new(
                    ModulePath::new(Package::new("m", ""), vec![]),
                    Default::default(),
                    vec![(
                        "Error".into(),
                        types::Record::new("Error", Default::default(), SourceInformation::dummy())
                            .into(),
                    )]
                    .into_iter()
                    .collect(),
                    Default::default(),
                ),
                false,
            )],
            vec![],
            vec![],
            vec![],
        );
    }

    fn create_let_error_transformer() -> Arc<LetErrorTransformer> {
        let reference_type_resolver = ReferenceTypeResolver::new(&MODULE);
        let type_equality_checker = TypeEqualityChecker::new(reference_type_resolver.clone());
        let type_canonicalizer = TypeCanonicalizer::new(
            reference_type_resolver.clone(),
            type_equality_checker.clone(),
        );

        LetErrorTransformer::new(
            reference_type_resolver,
            type_equality_checker,
            type_canonicalizer,
            ERROR_TYPE_CONFIGURATION.clone(),
        )
    }

    #[test]
    fn transform() {
        let error_type = types::Reference::new("Error", SourceInformation::dummy());

        insta::assert_debug_snapshot!(create_let_error_transformer().transform(
            &LetError::with_type(
                types::Union::new(
                    vec![
                        types::None::new(SourceInformation::dummy()).into(),
                        error_type.into()
                    ],
                    SourceInformation::dummy()
                ),
                vec![VariableDefinition::new(
                    "x",
                    Variable::new("y", SourceInformation::dummy()),
                    types::Number::new(SourceInformation::dummy()),
                    SourceInformation::dummy(),
                )],
                None::new(SourceInformation::dummy()),
                SourceInformation::dummy(),
            )
        ));
    }
}
