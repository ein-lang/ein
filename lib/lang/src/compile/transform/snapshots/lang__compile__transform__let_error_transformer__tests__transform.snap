---
source: lib/lang/src/compile/transform/let_error_transformer.rs
expression: "create_let_error_transformer().transform(&LetError::with_type(types::Union::new(vec![types\n                                                                                     ::\n                                                                                     None\n                                                                                     ::\n                                                                                     new(SourceInformation\n                                                                                         ::\n                                                                                         dummy())\n                                                                                     .\n                                                                                     into(),\n                                                                                     error_type\n                                                                                     .\n                                                                                     clone()\n                                                                                     .\n                                                                                     into()],\n                                                                                SourceInformation::dummy()),\n                                                              vec![VariableDefinition\n                                                                   ::\n                                                                   new(\"x\",\n                                                                       Variable\n                                                                       ::\n                                                                       new(\"y\",\n                                                                           SourceInformation\n                                                                           ::\n                                                                           dummy()),\n                                                                       types\n                                                                       ::\n                                                                       Number\n                                                                       ::\n                                                                       new(SourceInformation\n                                                                           ::\n                                                                           dummy()),\n                                                                       SourceInformation\n                                                                       ::\n                                                                       dummy(),)],\n                                                              None::new(SourceInformation::dummy()),\n                                                              SourceInformation::dummy()))"
---
Ok(
    Case(
        Case {
            type_: Union(
                Union {
                    types: {
                        Number(
                            Number {
                                source_information: SourceInformation {
                                    source_name: "",
                                    location: Location {
                                        line_number: 0,
                                        column_number: 0,
                                    },
                                    line: "",
                                },
                            },
                        ),
                        Reference(
                            Reference {
                                name: "Error",
                                source_information: SourceInformation {
                                    source_name: "",
                                    location: Location {
                                        line_number: 0,
                                        column_number: 0,
                                    },
                                    line: "",
                                },
                            },
                        ),
                    },
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            name: "x",
            argument: Variable(
                Variable {
                    name: "y",
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            alternatives: [
                Alternative {
                    type_: Number(
                        Number {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    expression: None(
                        None {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                },
                Alternative {
                    type_: Reference(
                        Reference {
                            name: "Error",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    expression: TypeCoercion(
                        TypeCoercion {
                            argument: Variable(
                                Variable {
                                    name: "x",
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            from: Reference(
                                Reference {
                                    name: "Error",
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            to: Union(
                                Union {
                                    types: {
                                        None(
                                            None {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        Reference(
                                            Reference {
                                                name: "Error",
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                    },
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                },
            ],
            source_information: SourceInformation {
                source_name: "",
                location: Location {
                    line_number: 0,
                    column_number: 0,
                },
                line: "",
            },
        },
    ),
)
